[ADR] Sending emails via api (#12054)
Sylius/Sylius: adr/2020_11_18_sending_emails_via_api.md

Date: 2020-11-18

# API - Sending emails

In power from version: master (?)

Related PRs: #11754, #12050

### Problem to solve

> How to send email using api

Reviewer: ```suggestion
How to send email using api.
```

### Context

> With more and more api coverage we had to decide how we should handle sending emails.

Reviewer: some operations in Sylius as the final step send emails to the customer, for full API coverage we need to handle it also by API

Reviewer: or sth like this :D

> ### Possible solutions

Reviewer: You described only one solution, so this title is misleading.

Using events allow us to trigger any handler once operation is completed. 

* Sylius\ApiBundle\Event\OrderCompleted

Once this event is triggered we handle it with ```src/Sylius/Bundle/ApiBundle/EventHandler/OrderCompletedHandler.php```

Then command ```src/Sylius/Bundle/ApiBundle/Command/SendOrderConfirmation.php ``` is dispatched, which triggers command handler (```src/Sylius/Bundle/ApiBundle/CommandHandler/SendOrderConfirmationHandler.php```)
 
SendOrderConfirmationHandler is sending email based on its destination (in this case OrderConfirmation)

```
$this->emailSender->send(
    Emails::ORDER_CONFIRMATION_RESENT,
    [$order->getCustomer()->getEmail()],
    [
        'order' => $order,
        'channel' => $order->getChannel(),
        'localeCode' => $order->getLocaleCode(),
    ]
);
```

### Decision and reasoning

> This adds alot of new abstraction, but using events we can trigger any email handler without digging into code. Only thing we need to do is handle triggered event (or create one if none is dispatched), send it to emailSender and we are good to go.

Reviewer: ```suggestion
This adds a lot of new abstraction, but using events we can trigger any email handler without digging into code. Only thing we need to do is handle triggered event (or create one if none is dispatched), send it to emailSender and we are good to go.
```

# API - Sending emails

* Status: proposed
* Date: 2020-11-18

## Context and Problem Statement

> To achieve 100% api coverage, we need to handle emails by api.

Reviewer 2: ```suggestion
To achieve 100% API coverage, we need to handle emails by API.
```

## Considered Options

### Using events

Using events allow us to trigger any handler once operation is completed. 

* Sylius\ApiBundle\Event\OrderCompleted

Once this event is triggered we handle it with ```src/Sylius/Bundle/ApiBundle/EventHandler/OrderCompletedHandler.php```

Then command ```src/Sylius/Bundle/ApiBundle/Command/SendOrderConfirmation.php ``` is dispatched, which triggers command handler (```src/Sylius/Bundle/ApiBundle/CommandHandler/SendOrderConfirmationHandler.php```)
 
SendOrderConfirmationHandler is sending email based on its destination (in this case OrderConfirmation)

```
$this->emailSender->send(
    Emails::ORDER_CONFIRMATION_RESENT,
    [$order->getCustomer()->getEmail()],
    [
        'order' => $order,
        'channel' => $order->getChannel(),
        'localeCode' => $order->getLocaleCode(),
    ]
);
```

> * Good, because its sending email based on triggered event, once understood it's simple to add new handlers.

Reviewer 3: ```suggestion
* Good, because it's easy to integrate other logic after the initial command will succeed.
* Good, because it's easy to trigger the same operation several times.
* Good, because it's easier to handle mailing server downs gracefully
```

# API - Sending emails

> * Status: proposed

Reviewer 4: @SirDomin @GSadee 

Shouldn't the status be "accepted" now?
